feat: Add ListMetadata + Read + Write functionality to Attio connector

sanjay-a01 opened the pull request:
Response:
![image](https://github.com/user-attachments/assets/6ef6dd02-292e-4ca8-8eb0-2598dfdcb2cd)

Cobalt0s commented on providers/attio/metadata.go:
```diff
@@ -0,0 +1,118 @@
+	}
+
+	fieldsMap := map[string]string{}
+	for k := range responseMap {
```
Since we are only collecting "keys", `ajson` library has method that will avoid additional marshall, unmarshall. It would look like this:
```
responseMap, err := body.GetObject()
```
https://github.com/spyzhov/ajson/blob/0d5fdeed023cf269f55ec781a2713153e3f3ca33/node.go#L494


Alternatively, if you want to deal with byte data directly avoiding ajson library, you could use the same UnmarshalJSON you use above but for map type:
```
responseMap, err := common.UnmarshalJSON[map[string]any](resp)
```

Cobalt0s replied:
You are serving Metadata from API and so far we were adding static `schema.json` for those providers that do not have such APIs. Therefore, no action is needed on this one.

.

**PS**: you can check out pipeliner to see how `schema.json` is created and served.
There is OpenAPI yaml or json commited under:
https://github.com/amp-labs/connectors/tree/main/providers/pipeliner/openapi
The `FileManager` loads it (file.go) and is referenced by 2 places:
* Schemas are objects object representation of that file and is used by `ListObjectMetadata` https://github.com/amp-labs/connectors/blob/main/providers/pipeliner/metadata.go.
* And most importantly this is the script which will differ among providers: https://github.com/amp-labs/connectors/blob/main/scripts/openapi/pipeliner/metadata/main.go. It loads OpenAPI file, processes it in custom way and outputs the `schema.json`.

In rare cases `schema.json` is created manually or a website is scrapped, which should be avoided for maintenance cost: https://github.com/amp-labs/connectors/tree/main/scripts/scrapper/salesloft/metadata

Here are some READMEs that I wrote:
* https://github.com/amp-labs/connectors/tree/main/scripts/openapi
* https://github.com/amp-labs/connectors/tree/main/tools/fileconv/api3

Cobalt0s commented on providers/attio/metadata.go:
```diff
@@ -0,0 +1,118 @@
+	// Other fields
+}
+
+var errCannotLoadMetadata = errors.New("cannot load metadata")
```
There is a recent `common` error that could fit for use case: `ErrMissingExpectedValues`.
Otherwise, export this error, and maybe even not within connector but from `common` package.

Cobalt0s commented on test/attio/metadata/metadata.go:
```diff
@@ -0,0 +1,24 @@
+
+	m, err := conn.ListObjectMetadata(ctx, []string{"objects", "lists", "self", "workspace_members", "webhooks", "tasks"})
+	if err != nil {
+		log.Fatal(err)
```
I was told to use `slog` lib and avoid `log`.
There is `utils.Fail()` that does that underneath.

Cobalt0s reviewed (commented):
Hi :wave:, some minor comments

sanjay-a01 commented:
updated unmarshal logic, used slog for error handling and common package vars

Cobalt0s reviewed (approved):
Everything is great! Some minor follow up comments.

Additionally, could you add mock tests in this or other PR.
For that, please record sample outputs of `c.Client.Get(ctx, url.String())` from Attio including successful response and some kind of failure. Mock tests would provide freedom to make changes to any code knowing that it will have our back and ensure that it functions as you intended.

sanjay-a01 commented:
- Removed code related to the self object ( Not supported as it is a single object and related to Auth info)
- Added mock test for a list of getting metadata.
- Implemented read functionality in the Attio connector

Cobalt0s commented on providers/attio/metadata_test.go:
```diff
@@ -0,0 +1,164 @@
+			ExpectedErrs: []error{common.ErrMissingObjects},
+		},
+		{
+			Name:  "Successfully describe multiple object with metadata",
```
This is a great test!

Cobalt0s commented on providers/attio/objectNames.go:
```diff
@@ -0,0 +1,23 @@
+)
+
+var supportedObjectsByRead = handy.NewSet( //nolint:gochecknoglobals
+	// Object Name	----------	API endpoint path
```
comment is redundant

Cobalt0s commented on providers/attio/parse.go:
```diff
@@ -0,0 +1,24 @@
+// recordsWrapperFunc returns the records using the objectName dynamically.
+func recordsWrapperFunc(obj string) common.RecordsFunc {
+	return func(node *ajson.Node) ([]map[string]any, error) {
+		result, err := jsonquery.New(node).Array(obj, true)
```
This method is similar to `common.GetRecordsUnderJSONPath()`. The only difference is that your version is ok with array being null as indicated by the `true` flag.

Maybe there should be 2 methods under common:
* GetRecordsUnderJSONPath -> which would be your version
* MustRecordsUnderJSONPath -> which is current impl GetRecordsUnderJSONPath.

Frankly, I am not sure that distinction is important!
I think before when I was refactoring, if key was not found that indicated that our code is at fault having wrong expectations in payload, thats about it... I have a feeling that you could change `GetRecordsUnderJSONPath` to have a `true` flag and use that method.

Cobalt0s commented on providers/attio/read.go:
```diff
@@ -0,0 +1,42 @@
+
+	return common.ParseResult(
+		rsp,
+		recordsWrapperFunc("data"),
```
As suggested by above, something along the lines of `common.GetRecordsUnderJSONPath("data")`

Cobalt0s reviewed (approved):
Looks awesome! Thanks for adding so many mock responses, this is good.

@RajatPawar please have a look

sanjay-a01 commented:
Used common functionality for unmarshalling the read response data

Cobalt0s commented on providers/attio/write.go:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	if len(resp.Data) == 0 {
+		return nil, ErrEmptyResultResponse
```
I think we can return success but with data as nil.

@RajatPawar what do you think about this suggestion? Also, when I was going over deep connectors I started wondering can common.WriteResult just have Success field?

RajatPawar replied:
Looking at other connectors, like @Cobalt0s said, this should probably be an empty but successful `WriteResult`

Cobalt0s reviewed (approved):
Write functionality looks good to me!

The PR title should be renamed.

Also it would be nice to have multiple dedicate PRs. Write and ListObjectMetadata are fully independent from Read.

RajatPawar commented on providers/attio/connector.go:
```diff
@@ -0,0 +1,63 @@
+)
+
+const (
+	apiVersion = "/v2"
```
Can you truncate the `/` out and use a method like filepath.Join (or if we have some other util function) instead wherever we use it? This is easy to make changes and cause bugs with

Cobalt0s replied:
`urlbuilder` deals with this issues. It is prefered to not have slashes for sure for readability.
```
// Constructor handles path
urlbuilder.New(c.BaseURL, uriPart1, uriPart2, uriPart3, uriPart4)
```

RajatPawar commented on providers/attio/connector.go:
```diff
@@ -0,0 +1,63 @@
+	}
+
+	// Read provider info
+	providerInfo, err := providers.ReadInfo(providers.Attio)
```
Can you use `Provider()` wherever we want Attio? Not a big problem right now, but it'll help in realizing common code paths and generalizing them

RajatPawar commented on providers/attio/parse.go:
```diff
@@ -0,0 +1,10 @@
+	"github.com/spyzhov/ajson"
+)
+
+var dummyNextPageFunc = func(*ajson.Node) (string, error) {
```
Can you move this over to `read`? Would be good to have lesser files whenever possible

Cobalt0s replied:
This function should belong in common. I can see other connectors using it. As a default behaviour -- no pagination.

sanjay-a01 replied:
I removed dummyNextPageFunc function because some object in Attio connector has pagination. Instead of this function, I created new function makeNextRecordsURL function in attio/parse.go file.

RajatPawar commented on providers/attio/types.go:
```diff
@@ -0,0 +1,11 @@
+
+var ErrEmptyResultResponse = errors.New("writing reponded with an empty result")
+
+type writeResponse struct {
```
I'd move all these to `write.go` for the same rationale as my other comment, especially since they also are all about writing

RajatPawar commented on providers/attio/write.go:
```diff
@@ -0,0 +1,58 @@
+	"github.com/amp-labs/connectors/common"
+)
+
+// Write creates/updates records in marketo. Write currently supports operations to the leads API only.
```
```suggestion
// Write creates/updates records in attio. Write currently supports operations to the leads API only.
```

RajatPawar commented on providers/attio/metadata.go:
```diff
@@ -0,0 +1,77 @@
+
+	for _, obj := range objectNames {
+		// Constructing the request url.
+		url, err := c.getApiURL(obj)
```
If there's a way to limit the response to just 1 object, that'd be great. We can avoid waiting on a larger response size

Cobalt0s replied:
I think I saw a code somewhere that you wrote that does gorotuines, generalizing it and amplying here would be nice.
Requesting in multithreaded fashion.

RajatPawar commented on providers/attio/README.md:
```diff
@@ -0,0 +1,16 @@
+
+Attio API version : v2
+
+| Object | Resource | Method
```
Write seems to be missing as of now?

Cobalt0s replied:
I think objectNames.go is an exact representation of this README. I would remove README, maybe would add comments to that file instead. There is always an incentive to keep go code up to date unlike `.md`, `.txt` type of files.

RajatPawar commented on providers/attio/README.md:
```diff
@@ -0,0 +1,16 @@
+
+
+## Supported Objects 
+Below is an exhaustive list of the supported Objects in the Attio deep connector with their endpoint resources(ObjectName).
```
```suggestion
Below is an exhaustive list of objects & methods supported on the objects -
```

RajatPawar reviewed (approved):
Approving this, please - 
1. Reply/resolve all the comments before merging it.
2. Attach screenshots of all the tests you've added under a 'Testing' section in the PR description
3. Wherever the test does not actually make calls to the API, please attach a screenshot of a live read/write/metadata call.

Other than that, the PR looks great! Congrats on getting your first connector in 🎉

sanjay-a01 commented:
### Getting a List of Metadata for the supporting objects
![image](https://github.com/user-attachments/assets/ae8a30e5-27be-4080-8cab-4e7f978e7c80)

### Test Result for list of Metadata
![image](https://github.com/user-attachments/assets/8d90fab3-b2ae-4c56-97ea-ef5a1a05a2ec)

### Read Functionality Responses
![image](https://github.com/user-attachments/assets/c04ee327-2b0d-49ac-86da-7718cb715b4e)
![image](https://github.com/user-attachments/assets/1b7ebb3d-78dd-4ab6-8c1d-bff693d6ec9f)
![image](https://github.com/user-attachments/assets/35994d55-e44b-4176-a00c-fc025e7e3971)
![image](https://github.com/user-attachments/assets/f65f676b-7085-4865-826d-a81aadb536ba)

### Write Functionality Responses
![image](https://github.com/user-attachments/assets/cf6caea9-61d4-415b-a074-9270a71da44d)
![image](https://github.com/user-attachments/assets/7526c322-7e1a-4693-ace3-1f38e7daf8ae)
![image](https://github.com/user-attachments/assets/af5d55b3-850f-47e5-8b41-80473ed413d8)
![image](https://github.com/user-attachments/assets/73aa92c3-4dfe-4442-aacc-b341d7269a9b)

### About creating Webhooks
"Target Url" is one of the body parameters used to create the webhook.
To make the target URL, I created a gin server and implemented the POST method to handle the URL.  Then using ngrok to expose my local server and get the public URL.
With the use of the URL, we can create the webhook. While the Event is triggered, the webhook URL automatically triggered by Attio server.
Also, we can monitor the changes in the ngrok dashboard page.

### Adding Query parameter in read functionality
Some supporting objects have pagination in the Attio connector but, we give the pagination value through a query parameter. So I added the url.WithQueryParams() function manually in the read functionality. If the object has a process to handle the query parameter then the query value will be added to the URL or otherwise, it will neglected.

Jkarage commented on providers/attio/metadata.go:
```diff
@@ -0,0 +1,77 @@
+// nolint
```
Hey @sanjay-a01 ,Is there a reason for disabling linter in these files: metadata, ObjectNames, read, write?

sanjay-a01 replied:
Yeah, I faced the lint issue while raising the PR so only I have disabled the linter.

Cobalt0s replied:
Yeah, I dont think it should be disabled entirely. Some things are to be addressed and some should be ignored line by line. Otherwise, if more inconsistencies are added the linter would not report them, which is not good.

Jkarage commented on providers/attio/metadata.go:
```diff
@@ -0,0 +1,77 @@
+
+// This struct is used for when the response data having slice of data.
+type responseObject struct {
+	Result []map[string]any `json:"data"`
```
I would have named this Data instead of Result.

sanjay-a01 replied:
If we specify the Result instead of Data, when we are unmarshalling the response for getting metadata it can't unmarshal because the response data contains "data" as the key.
![image](https://github.com/user-attachments/assets/bdaa7d23-7bf1-444e-8708-35ad534dfae1)

Cobalt0s replied:
I think the suggestion is 
```
Data []map[string]any `json:"data"`
```
if understood correctly.

Jkarage commented on providers/attio/write.go:
```diff
@@ -0,0 +1,63 @@
+		// that we are extending 'List' resource and creating a new record
+		write = c.Client.Post
+	} else {
+		// only put is supported for updating 'Single' resource
```
```suggestion
		// uses patch in updating 'single' resource.
```

Jkarage commented on providers/attio/write.go:
```diff
@@ -0,0 +1,63 @@
+	Data    map[string]any `json:"data"`
+}
+
+// Write creates/updates records in attio. Write currently supports operations to the leads API only.
```
```suggestion
// Write creates/updates records in attio.
```

Cobalt0s commented on providers/attio/read.go:
```diff
@@ -0,0 +1,44 @@
+		return nil, err
+	}
+
+	url.WithQueryParam("limit", "1")
```
This doesn't seem right. This is correct during development process. We don't want to query only 1 item at a time!
Also if some objects have offset pagination support that should be implemented instead of `dummyNextPageFunc`.

Cobalt0s reviewed (changes requested):
There is an action needed regarding `read.go`. You need to handle pagination or at least remove `per page limit of 1`.

sanjay-a01 commented:
I have changed the pagination logic for the read functionality in the Attio connector

![image](https://github.com/user-attachments/assets/ae6de0cb-570f-4a0a-b449-369ffb4137bc)

In that while using **DumpJSON**  function for printing the response, I faced issue with the next page response like 

![image](https://github.com/user-attachments/assets/28e4ecbe-181a-4849-8d76-d00c80f2f55b)

and also I tried with **Json.MarshalIndent** function again getting the same result.
so I changed the marshalIndent logic like below

![image](https://github.com/user-attachments/assets/d09e92a5-cabb-418d-945c-dc688484b24f)
![image](https://github.com/user-attachments/assets/3a258276-995c-4310-98b8-a42c842638a2)

and got the nextpage URL like

![image](https://github.com/user-attachments/assets/adbc4635-87f0-4093-982b-4956a1536b3e)

It is just printing the response so I removed my marshalIndent logic and used **json.MarshalIndent** function for printing the response.

some object in the Attio connector don't require pagination, so I make the restriction using supportLimitAndOffset variable and check in the read functionality like

![image](https://github.com/user-attachments/assets/499d6344-28ac-4b7f-abe3-fe99a3b117b9)
![image](https://github.com/user-attachments/assets/8c01229c-d37e-4ee4-a46e-15cd4de473c7)

If the URL contains query parameters like limit and offset and has a response value, then only next page data will generate in the response otherwise it will return empty value in the next page

![image](https://github.com/user-attachments/assets/fd88e807-569c-415f-a92b-e2a5c11f94b9)

Cobalt0s commented on providers/attio/parse.go:
```diff
@@ -0,0 +1,43 @@
+		// Extract the data key value from the response.
+		value := node.MustObject()["data"].MustArray()
+
+		if (url.Query().Has("limit") || url.Query().Has("offset")) && len(value) != 0 {
```
I think it is worth changing urlbuilder.URL to address this issue instead of converting to url.URL.
urlbuilder holds query parameters already:
```
type URL struct {
	delegate           *url.URL
	queryParams        url.Values
	encodingExceptions map[string]string
}
```
Just add `HasQueryParam(name string)` method and delegate the call `u.queryParams.Has(name)`.

Cobalt0s commented on providers/attio/parse.go:
```diff
@@ -0,0 +1,43 @@
+			reqLink.WithQueryParam("offset", strconv.Itoa(nextStart))
+
+			return reqLink.String(), nil
+		} else {
```
```suggestion
		}
		
		return "", nil
```
Else keyword is not needed due to `return` under if statement.

Cobalt0s reviewed (changes requested):
Everything is great! I see lots of effort and well done!

One important thing is check the comments for `makeNextRecordsURL` method.
You should update `urlbuilder` package. And make use of `jsonquery` to avoid `panic()` calls.

sanjay-a01 commented on providers/attio/params.go:
```diff
@@ -0,0 +1,42 @@
+
+const (
+	// DefaultPageSize is number of elements per page.
+	DefaultPageSize = 10
```
yes, I have added the default size which is mentioned in the attio documentation

Cobalt0s reviewed (approved):
Awesome!
LGTM
